allennlp/data/dataset_readers/visual_entailment.py

```diff
@@ -0,0 +1,98 @@
+import logging
+
+from typing import (
+    Dict,
+    List,
+    Union,
+    Optional,
+    Tuple,
+)
+
+from overrides import overrides
+from torch import Tensor
+
+from allennlp.data.dataset_readers.dataset_reader import DatasetReader
+from allennlp.data.fields import ArrayField, LabelField, TextField
+from allennlp.data.instance import Instance
+from allennlp.data.dataset_readers.vision_reader import VisionReader
+
+from allennlp.common.file_utils import json_lines_from_file
+
+logger = logging.getLogger(__name__)
+
+
+@DatasetReader.register("visual-entailment")
+class VisualEntailmentReader(VisionReader):
+    """
+    The dataset reader for visual entailment.
+    """
+
+    @overrides
+    def _read(self, file_path: str):
+        lines = json_lines_from_file(file_path)
+        info_dicts: List[Dict] = list(self.shard_iterable(lines))  # type: ignore
+
+        if not self.skip_image_feature_extraction:
+            # It would be much easier to just process one image at a time, but it's faster to process
+            # them in batches. So this code gathers up instances until it has enough to fill up a batch
+            # that needs processing, and then processes them all.
+            processed_images = self._process_image_paths(
+                [self.images[info_dict["Flickr30K_ID"] + ".jpg"] for info_dict in info_dicts]
+            )
+        else:
+            processed_images = [None for i in range(len(info_dicts))]  # type: ignore
+
+        attempted_instances_count = 0
+        failed_instances_count = 0
+        for info_dict, processed_image in zip(info_dicts, processed_images):
+            hypothesis = info_dict["sentence2"]
+            answer = info_dict["gold_label"]
+
+            instance = self.text_to_instance(processed_image, hypothesis, answer)
+            attempted_instances_count += 1
+            if instance is None:
+                failed_instances_count += 1
+            else:
+                yield instance
+
+            if attempted_instances_count % 2000 == 0:
+                failed_instances_fraction = failed_instances_count / attempted_instances_count
+                if failed_instances_fraction > 0.1:
+                    logger.warning(
+                        f"{failed_instances_fraction*100:.0f}% of instances have no answers."
+                    )
```

Comment:
Now, `text_to_instance()` will never return `None` anymore, so we don't need this check either.